### PR TITLE
Order Creation: Add discard changes prompt to new order customer details

### LINKED ISSUE #6454: Order Creation: Add "discard changes" prompt to customer details screen
Order editing includes a "discard changes" prompt when you swipe to close the customer details modal on an existing order (see https://github.com/woocommerce/woocommerce-ios/issues/6449). This protects against losing data you've entered if you accidentally swipe to close the modal view.

I think this would be a nice enhancement to the customer details modal in order creation, too. It would be frustrating to enter data there and then lose it if you accidentally swipe it closed while you navigate around the form.

### PR DESCRIPTION
**Pull request by rachelmcr**
Closes: #6454 

## Description

This adds a "discard changes" prompt when swiping to dismiss the Customer Details modal on the New Order screen, to prevent accidental data loss. The prompt only appears when there are unsaved changes.

The prompt is added with a SwiftUI view modifier, which should be reusable with other SwiftUI modal views in the future.

## Changes

* Adds a SwiftUI view modifier `discardChangesPrompt(canDismiss:didDismiss:)`:
   * This modifier wraps the SwiftUI in a UIHostingController (in `DiscardChangesWrapper`), which uses `canDismiss` to determine if the modal can be dismissed with a swipe gesture and fires `didDismiss` when the modal is dismissed.
   * If `canDismiss` is false, it uses our `UIAlertController.presentDiscardChangesActionSheet` helper to display the discard changes prompt.
* Moves `hasPendingChanges()` from the `EditOrderAddressFormViewModel` (used in order editing) to `AddressFormViewModelProtocol` so it can also be used for order creation to determine if the address form has unsaved changes.
* Adds a manually created binding `canDismiss` to `CreateOrderAddressFormViewModel`. I couldn't find a better way to create this binding to use with the new view modifier; let me know if you have an alternate suggestion for this.
* Adds the view modifier to the modal in `OrderCustomerSection`, using the new binding to determine if the modal can be dismissed without the prompt.

## Testing

1. Go to the Orders tab and create a new order.
2. Select "Add Customer Details."
3. Confirm you can close the modal using the "Close" button and swiping to dismiss it before you've made any changes.
4. Add some customer details and try swiping to dismiss the modal. Confirm a "discard changes" prompt appears and works as expected.
5. Add some customer details and confirm you can tap the "Done" button to save your changes.

## Screenshots

https://user-images.githubusercontent.com/8658164/160631566-7f8e2dff-814e-4e3a-b100-a17a2fca3687.mp4

## Submitter Checklist

Update release notes:
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

File: WooCommerce/Classes/View Modifiers/View+DiscardChanges.swift [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+        init(_ wrapper: DiscardChangesWrapper) {
+            self.wrapper = wrapper
+        }
+
+        func presentationControllerShouldDismiss(_ presentationController: UIPresentationController) -> Bool {
+            wrapper.canDismiss
+        }
+
+        func presentationControllerDidAttemptToDismiss(_ presentationController: UIPresentationController) {
+            UIAlertController.presentDiscardChangesActionSheet(viewController: presentationController.presentedViewController) { [weak self] in
```
Comment (Ecarrion) [inline WooCommerce/Classes/View Modifiers/View+DiscardChanges.swift]: I wonder if we could inject somehow this action sheet, I could see cases where we don't want this specific message, what do you think?

**Review by Ecarrion — CHANGES_REQUESTED** [review]
Hi @rachelmcr, 

Your `@Binding` question made me realize that I think we are interpreting the `SwiftUI` lifecycle in a slightly wrong way.

Basically, we need to treat the wrapper view as a disposable view, and update the underlying coordinator(state manager) on each update pass.

This implies that we need to hold the `addressFormViewModel` in some way, ideally in the `newOrderViewModel`.

As these changes are quite scattered(and I needed to validate what I'm saying 😛) I have used some rough changes for you to take a look at [in this diff](https://github.com/woocommerce/woocommerce-ios/compare/issue/6465-customer-details-discard-changes...issue/discard-proposed-changes?expand=1)

Let me know what you think!

**Comment by rachelmcr** [comment]
@Ecarrion Thank you! I suspected I was missing something there, and that the `@Binding` wasn't quite right since we only needed to read the value, but I only got myself part of the way to the solution you worked out. (And partway there was never enough for it to work right, hehe.)

From a quick look at your changes I think this makes a lot of sense. 👍  I'll look at them more closely and work them into this PR, along with the other enhancement you suggested.

**Comment by rachelmcr** [comment]
@Ecarrion This is ready for another look! I incorporated your proposed changes in [87aa140](https://github.com/woocommerce/woocommerce-ios/pull/6534/commits/87aa140794602a82ba8e4c483b0ab9f40c89250f), holding the `addressFormViewModel` in the `newOrderViewModel`. It is passed it from there to the `OrderCustomerSection`, where it is an observed object (to make sure the `hasPendingChanges` value is observed correctly). When the address form is closed, a trigger is fired to reset the view model to make sure changes are discarded.

I also followed your suggestion to inject the discard changes action sheet in [bc496ea](https://github.com/woocommerce/woocommerce-ios/pull/6534/commits/bc496ea54a6b3cb7bca54adc03805d66d09decb7), to make that view modifier more flexible.

File: WooCommerce/Classes/View Modifiers/View+DiscardChanges.swift [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+
+    /// Optional method to be invoked when the view is dismissed.
+    ///
+    let didDismiss: (() -> Void)?
+
+    typealias UIViewControllerType = UIHostingController<Content>
+
+    func makeUIViewController(context: Context) -> UIViewControllerType {
+        let viewController = UIHostingController(rootView: view)
+        return viewController
```
Comment (Ecarrion) [inline WooCommerce/Classes/View Modifiers/View+DiscardChanges.swift]: super nit 😅 
```suggestion
        UIHostingController(rootView: view)
```

File: WooCommerce/Classes/View Modifiers/View+DiscardChanges.swift [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+    func updateUIViewController(_ uiViewController: UIViewControllerType, context: Context) {
+        context.coordinator.wrapper = self
+        uiViewController.parent?.presentationController?.delegate = context.coordinator
+    }
+
+    func makeCoordinator() -> Self.Coordinator {
+        Coordinator(self)
+    }
+
+    class Coordinator: NSObject, UIAdaptivePresentationControllerDelegate {
```
Comment (Ecarrion) [inline WooCommerce/Classes/View Modifiers/View+DiscardChanges.swift:47]:
```suggestion
    private class Coordinator: NSObject, UIAdaptivePresentationControllerDelegate {
```

File: WooCommerce/Classes/ViewRelated/Orders/Order Creation/CustomerSection/OrderCustomerSection.swift [hunk 4]
```diff
@@ -7,22 +7,27 @@ struct OrderCustomerSection: View {
...
+                NavigationView {
+                    EditOrderAddressForm(dismiss: {
+                        showAddressForm.toggle()
+                    }, viewModel: addressFormViewModel)
+                }
+                .discardChangesPrompt(presentActionSheet: addressFormViewModel.presentDiscardChangesActionSheet,
+                                      canDismiss: !addressFormViewModel.hasPendingChanges,
+                                      didDismiss: addressFormViewModel.userDidCancelFlow)
+                .onDisappear {
+                    viewModel.resetAddressFormTrigger.send()
```
Comment (Ecarrion) [inline WooCommerce/Classes/ViewRelated/Orders/Order Creation/CustomerSection/OrderCustomerSection.swift]: Could you explain a little bit the reason about why you preferred a `Combine` publisher over of a regular method?

File: WooCommerce/Classes/ViewRelated/Orders/Order Creation/NewOrder.swift [hunk 5]
```diff
@@ -98,7 +98,8 @@ struct NewOrder: View {
 
                         Spacer(minLength: Layout.sectionSpacing)
 
-                        OrderCustomerSection(viewModel: viewModel)
+                        OrderCustomerSection(viewModel: viewModel,
+                                             addressFormViewModel: viewModel.addressFormViewModel ?? viewModel.createOrderAddressFormViewModel())
```
Comment (Ecarrion) [inline WooCommerce/Classes/ViewRelated/Orders/Order Creation/NewOrder.swift]: Do you think we should wrap this inside some kind of `if let`? From looking at the code it seems like the  `viewModel.createOrderAddressFormViewModel()` part will never be needed but It's misleading to the reader.

File: WooCommerce/Classes/ViewRelated/Orders/Order Creation/CustomerSection/CreateOrderAddressFormViewModel.swift [hunk 6]
```diff
@@ -36,6 +37,15 @@ final class CreateOrderAddressFormViewModel: AddressFormViewModel, AddressFormVi
...
     }
 
+    /// Creates and presents a discard changes action sheet.
+    ///
+    func presentDiscardChangesActionSheet(viewController: UIViewController) {
+        UIAlertController.presentDiscardChangesActionSheet(viewController: viewController) { [weak self] in
+            viewController.dismiss(animated: true, completion: nil)
+            self?.userDidCancelFlow()
+        }
+    }
```
Comment (Ecarrion) [inline WooCommerce/Classes/ViewRelated/Orders/Order Creation/CustomerSection/CreateOrderAddressFormViewModel.swift]: Seeing the `ViewModel` having to `import UIKit` raised my alarm 😆 

I think we should keep the `UIKit` bit constrained to the `DiscardChangesWrapper`, what do you think about just allowing the consumer to provide the `title` and the `action button title`, so then we can call the UIKit action sheet inside the wrapper. something like:

```swift
func discardChangesPrompt(title: String,
                          discardButtonTitle: String,
                          canDismiss: Bool,
                          didDismiss: (() -> Void)? = nil) -> some View {
```

**Review by Ecarrion — COMMENTED** [review]
Looking much better!
Left you another suggestion to keep our view models clean from the `UIKit` domain!

File: WooCommerce/Classes/ViewRelated/Orders/Order Creation/CustomerSection/CreateOrderAddressFormViewModel.swift (hunk 6 above)
Comment (rachelmcr) [reply WooCommerce/Classes/ViewRelated/Orders/Order Creation/CustomerSection/CreateOrderAddressFormViewModel.swift]: That's fair. I was trying to make it flexible enough to handle action sheets like the one we use for product details, which has more actions (discard changes, save as draft, cancel). But we don't have a need for that yet, and if we get to that point we can always handle the complexity in a `UIAlertController` helper rather than importing UIKit li.

I'll go ahead and change it to something more like your suggestion 👍

File: WooCommerce/Classes/ViewRelated/Orders/Order Creation/CustomerSection/OrderCustomerSection.swift (hunk 4 above)
Comment (rachelmcr) [reply WooCommerce/Classes/ViewRelated/Orders/Order Creation/CustomerSection/OrderCustomerSection.swift]: 🤔  When I first set up that trigger, I thought the separate method `createOrderAddressFormViewModel()` wouldn't be needed and everything could be done within `configureAddressFormViewModel()` using `Combine`. That didn't work out, so now that I'm looking at this with fresh eyes I think this probably isn't needed and I could just simplify this line like this:

```
viewModel.addressFormViewModel = viewModel.createOrderAddressFormViewModel()
```

File: WooCommerce/Classes/View Modifiers/View+DiscardChanges.swift (hunk 3 above)
Comment (rachelmcr) [reply WooCommerce/Classes/View Modifiers/View+DiscardChanges.swift:47]: This can't be made private because it's part of the `UIViewControllerRepresentable` protocol that `DiscardChangesWrapper` conforms to.

File: WooCommerce/Classes/ViewRelated/Orders/Order Creation/CustomerSection/OrderCustomerSection.swift (hunk 4 above)
Comment (rachelmcr) [reply WooCommerce/Classes/ViewRelated/Orders/Order Creation/CustomerSection/OrderCustomerSection.swift]: Updated in [7626e15](https://github.com/woocommerce/woocommerce-ios/pull/6534/commits/7626e15d3db238d837e898ae6fb089818f755e3d).

File: WooCommerce/Classes/ViewRelated/Orders/Order Creation/NewOrder.swift (hunk 5 above)
Comment (rachelmcr) [reply WooCommerce/Classes/ViewRelated/Orders/Order Creation/NewOrder.swift]: Added this in [7626e15](https://github.com/woocommerce/woocommerce-ios/pull/6534/commits/7626e15d3db238d837e898ae6fb089818f755e3d), with documentation to clarify that it's just to handle the optional `viewModel.addressFormViewModel`.

File: WooCommerce/Classes/ViewRelated/Orders/Order Creation/CustomerSection/CreateOrderAddressFormViewModel.swift (hunk 6 above)
Comment (rachelmcr) [reply WooCommerce/Classes/ViewRelated/Orders/Order Creation/CustomerSection/CreateOrderAddressFormViewModel.swift]: Updated in [59c090e](https://github.com/woocommerce/woocommerce-ios/pull/6534/commits/59c090eb868f19e7c4e382adfc6cae2fd99e0f5a). I decided to provide a set of default strings (since I expect that in most cases where we have a discard changes action sheet we'll be ok with those defaults) with the option to override them.

**Comment by rachelmcr** [comment]
Thanks again for the feedback, @Ecarrion. This is ready for another look!

I'm still not entirely happy with the way the `CreateOrderAddressFormViewModel` view model is being configured (it feels a little hacky) but I couldn't think of a better way to do it, and at least this way is fairly simple/straightforward compared to earlier iterations.

File: WooCommerce/Classes/ViewRelated/Orders/Order Creation/CustomerSection/OrderCustomerSection.swift [hunk 7]
```diff
@@ -7,22 +7,26 @@ struct OrderCustomerSection: View {
...
+            .sheet(isPresented: $showAddressForm) {
+                NavigationView {
+                    EditOrderAddressForm(dismiss: {
+                        showAddressForm.toggle()
+                    }, viewModel: addressFormViewModel)
+                }
+                .discardChangesPrompt(canDismiss: !addressFormViewModel.hasPendingChanges,
+                                      didDismiss: addressFormViewModel.userDidCancelFlow)
+                .onDisappear {
+                    viewModel.addressFormViewModel = viewModel.createOrderAddressFormViewModel()
```
Comment (Ecarrion) [inline WooCommerce/Classes/ViewRelated/Orders/Order Creation/CustomerSection/OrderCustomerSection.swift]: One minor nit, I would have created a method called `viewModel.resetAddressForm()` to hide this complexity. Then we can add comments on why is this needed.

File: WooCommerce/Classes/ViewRelated/Orders/Order Creation/NewOrder.swift [hunk 8]
```diff
@@ -131,6 +131,16 @@ struct NewOrder: View {
...
+
+    /// View model for Customer section address form.
+    ///
+    private var addressFormViewModel: CreateOrderAddressFormViewModel {
+        guard let addressFormViewModel = viewModel.addressFormViewModel else {
+            // Returns a newly created view model if one isn't set. This shouldn't be needed but handles the optional CreateOrderAddressFormViewModel.
+            return viewModel.createOrderAddressFormViewModel()
+        }
+        return addressFormViewModel
+    }
```
Comment (Ecarrion) [inline WooCommerce/Classes/ViewRelated/Orders/Order Creation/NewOrder.swift]: I wonder if this should leave inside the `vm`. what do you think?

**Review by Ecarrion — APPROVED** [review]
Looks great!, just some minor nits but then 🚀

**Comment by rachelmcr** [comment]
@Ecarrion Could I trouble you to take one more look at this? Your comments above made me rethink the approach — in particular, why we're resetting the address form by recreating the view model in the `NewOrderViewModel`. It occurs to me that a more straightforward approach is to reset it (by resetting all the address fields) from within `CreateOrderAddressFormViewModel` as part of the cancel action.

I tried that in [c776816](https://github.com/woocommerce/woocommerce-ios/pull/6534/commits/c776816f19d80408d5b7a2311f34008d8a4c6b09) and it simplifies a lot of things, and I think makes it clearer what we're doing. What do you think?

File: WooCommerce/Classes/ViewRelated/Orders/Order Creation/NewOrderViewModel.swift [hunk 9]
```diff
@@ -130,6 +130,19 @@ final class NewOrderViewModel: ObservableObject {
...
+    var addressFormViewModel: CreateOrderAddressFormViewModel {
+        CreateOrderAddressFormViewModel(siteID: siteID,
+                                        addressData: .init(billingAddress: orderSynchronizer.order.billingAddress,
+                                                           shippingAddress: orderSynchronizer.order.shippingAddress),
+                                        onAddressUpdate: { [weak self] updatedAddressData in
+            let input = Self.createAddressesInput(from: updatedAddressData)
+            self?.orderSynchronizer.setAddresses.send(input)
+            self?.trackCustomerDetailsAdded()
+        })
+    }
```
Comment (Ecarrion) [inline WooCommerce/Classes/ViewRelated/Orders/Order Creation/NewOrderViewModel.swift]: I tested the feature and it seems to work but I can see that now this view model is not being retained, right?

Should we make this a lazy var to be on the safe side?

**Review by Ecarrion — COMMENTED** [review]
Looks good, I just have one concern with the lifecycle of the `addressFormViewModel`.

Also, I think we messed with the landscape constraints when we added the wrapper.
Screen Shot 2022-04-06 at 8 04 52 AM

File: WooCommerce/Classes/ViewRelated/Orders/Order Creation/NewOrderViewModel.swift (hunk 9 above)
Comment (rachelmcr) [reply WooCommerce/Classes/ViewRelated/Orders/Order Creation/NewOrderViewModel.swift]: Oh, you're right. But making this a lazy var doesn't work as-is because then what we pass in for its `addressData` isn't ever updated. Handling that takes us back to something more like how it was done in [7626e15](https://github.com/woocommerce/woocommerce-ios/pull/6534/commits/7626e15d3db238d837e898ae6fb089818f755e3d), so it seems I was mistaken about this being a better approach. 😬 

I think I'll revert these latest changes and address the comments from [your previous review](https://github.com/woocommerce/woocommerce-ios/pull/6534#pullrequestreview-932565888) + the UI issue you found here.

**Comment by rachelmcr** [comment]
@Ecarrion Ok, one more time. :) I reverted my latest changes to start from the approach used in [this previous review](https://github.com/woocommerce/woocommerce-ios/pull/6534#pullrequestreview-932565888). In [0296f26](https://github.com/woocommerce/woocommerce-ios/pull/6534/commits/0296f266594b5c2589c55136112c2e1782139958) I made a couple changes to it:

1. I created a `resetAddressForm` method that recreates the address form view model, and used that in `OrderCustomerSection`.
2. I made `NewOrderViewModel.addressFormViewModel` non-optional, to avoid that point of confusion. Instead, I initialize that property with a temporary view model first and then use `resetAddressForm` right away to reset it with the expected values. (I think this is less confusing than how it was done before, but let me know if you think it's better the other way.)

I also fixed the landscape issue — that was a weird one but it has a fairly simple fix, made in [992a409](https://github.com/woocommerce/woocommerce-ios/pull/6534/commits/992a409ed6d94f4d8f778d6433f245ebc44849e2).

Thanks for bearing with me while I wrap my head around the details of SwiftUI lifecycles!

**Review by Ecarrion — APPROVED** [review]
Looks & tests well! 🚀 

_PS: Sorry for dragging it for so long!_
